proof/compress-p1.saw

```diff
@@ -0,0 +1,201 @@
+  do {
+    (simplify (addsimps [mul_commutes_fp_thm, reciprocal_div_thm] empty_ss));
+    unfolding ["POINTonE1_affine_invariant", "POINTonE1_affine_rep"];
+    (apply_fp_rewrites);
+    w4_unint_z3 fp_unints;
+  };
+
+// POINTonE1_Compress_BE compresses X and returns the sign of Y.  It supports
+// both affine points (indicated by Z == 1) and projective points (Z != 1).
+let POINTonE1_Compress_BE_spec is_point_affine = do {
+  out_ptr <- crucible_alloc (llvm_array 48 (llvm_int 8));
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "in" POINTonE1_type;
+
+  crucible_precond {{ POINTonE1_invariant inp }};
+  if is_point_affine then do {
+    // Point is affine, so Z must be 1
+    crucible_precond {{ (fp_abs (inp.2)) == Fp.field_unit }};
+  } else do {
+    // Point is projective, so Z must not be 1
+    crucible_precond {{ (fp_abs (inp.2)) != Fp.field_unit }};
+    crucible_precond {{ ~(is_zero Fp ((POINTonE1_abs inp).2)) }};
+    crucible_precond {{ is_point_projective E (POINTonE1_abs inp) }}; // on the curve
+  };
+
+  crucible_execute_func [out_ptr, inp_ptr];
+
+  if is_point_affine then do {
+    crucible_points_to out_ptr (crucible_term {{ I2OSP48 (fp_abs inp.0) }});
+    crucible_return (crucible_term {{ (zext [sign_F_p (fp_abs inp.1), (HTCE1::sgn0 (fp_abs inp.1))@0]):[64] }});
+  } else do {
+    // Point is not already in affine coordinates, so must affinify first
+    let affinified = {{ affinify E (POINTonE1_abs inp) }};
+    crucible_points_to out_ptr (crucible_term {{ I2OSP48 (affinified.0) }});
+
+    // SAW struggles to reason about `HTCE1::sgn0 (affinified.1)`, but we don't
+    // need it anyway so this spec only specifies the sign_F_p bit.
+    ret <- crucible_fresh_var "ret" limb_type;
+    crucible_postcond {{ (ret!1) == (sign_F_p affinified.1) }};
+    crucible_return (crucible_term ret);
+  };
+};
+
+let POINTonE1_Compress_BE_proof = do {
+     // Unfolding needed for fp rewrites to apply
+     unfolding ["POINTonE1_affine_abs", "POINTonE1_affine_rep", "POINTonE1_invariant"];
+     (apply_fp_rewrites);
+     simplify (addsimp one_mont_p_abs_thm empty_ss);
+     w4_unint_z3 ["POINTonE1_abs", "POINTonE1_invariant", "Fp", "fp_abs", "I2OSP48"];
+  };
+
+// Prove POINTonE1_Compress_BE matches the spec for both projective and affine
+// points.
+POINTonE1_Compress_BE_projective_ov <- custom_verify "POINTonE1_Compress_BE"
+  [POINTonE1_from_Jacobian_ov, POINTonE1_affine_Compress_BE_ov, vec_is_equal_fp_ov]
+  (POINTonE1_Compress_BE_spec false)
+  (POINTonE1_Compress_BE_proof);
+POINTonE1_Compress_BE_affine_ov <- custom_verify "POINTonE1_Compress_BE"
+  [POINTonE1_affine_Compress_BE_ov, vec_is_equal_fp_ov]
+  (POINTonE1_Compress_BE_spec true)
+  (POINTonE1_Compress_BE_proof);
+
+
+let blst_p1_compress_spec is_point_affine = do {
+  out_ptr <- crucible_alloc (llvm_array 48 (llvm_int 8));
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "in" POINTonE1_type;
+  crucible_precond {{ POINTonE1_invariant inp }};
+
+  if is_point_affine then do {
```

Comment:
Nitpick: would it be possible to remove the distinction between Z=1 and Z!=1 in the spec? That would make things clearer. I understand that the code paths are different, and so this might ease the proof, but the spec for `is_point_affine=false` should work in both cases. Or did I miss something?